common/src/main/java/io/pravega/common/concurrent/Futures.java

```diff
@@ -543,6 +543,32 @@
         result.whenComplete((r, ex) -> sf.cancel(true));
         return result;
     }
+    
+    /**
+     * Creates a new CompletableFuture that either holds the result of future from the futureSupplier
+     * or will timeout after the given amount of time.
+     *
+     * @param futureSupplier  Supplier of the future. 
+     * @param timeout         The timeout for the future.
+     * @param executorService An ExecutorService that will be used to invoke the timeout on.
+     * @param <T>             The Type argument for the CompletableFuture to create.
+     * @return The result.
+     */
+    public static <T> CompletableFuture<T> futureWithTimeout(Supplier<CompletableFuture<T>> futureSupplier, 
+                                                             Duration timeout, ScheduledExecutorService executorService) {
+        CompletableFuture<T> result = new CompletableFuture<>();
```

Comment:
Is there a reason why we need to create a new CompletableFuture object? Can we avoid it ?

Reply:
i kept as separate future so that we can issue a cancel on the future received from futureSupplier. 
so at least any downstream processing that was chained may get cancelled.

Reply:
I have made the change to not create a new future. I have also added a method in futures to add timeout to an existing future.

Reply:
Thanks Shivesh.
cancel() does not cancel the computation as per the java doc.
``` 
/**
     * If not already completed, completes this CompletableFuture with
     * a {@link CancellationException}. Dependent CompletableFutures
     * that have not already completed will also complete
     * exceptionally, with a {@link CompletionException} caused by
     * this {@code CancellationException}.
     *
     * @param mayInterruptIfRunning this value has no effect in this
     * implementation because interrupts are not used to control
     * processing.
     *
     * @return {@code true} if this task is now cancelled
     */
    public boolean cancel(boolean mayInterruptIfRunning)
```